sbi/src/pmu.rs

```diff
@@ -0,0 +1,753 @@
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::error::*;
+use crate::function::*;
+
+/// Functions for the Performance Monitoring Unit (PMU) extension
+/// Specific details can be found in the SBI documentation for the PMU extension.
+#[derive(Copy, Clone)]
+pub enum PmuFunction {
+    /// Returns the total number of performance counters (hardware and firmware).
+    GetNumCounters,
+    /// Returns information about hardware counter specified by the inner value.
+    GetCounterInfo(u64),
+    /// Configures the counters selected by counter_index_base and counter_index_mask.
+    /// See the sbi_pmu_counter_config_matching documentation for details.
+    ConfigureMatchingCounters {
+        /// Counter index base.
+        counter_index_base: u64,
+        /// Counter index mask.
+        counter_index_mask: u64,
+        /// Counter configuration flags.
+        config_flags: PmuCounterConfigFlags,
+        /// Counter event type.
+        event_type: PmuEventType,
+        /// Counter event data.
+        event_data: u64,
+    },
+    /// Starts the counters selected by counter_index_base and counter_index_mask.
+    /// See the sbi_pmu_counter_start documentation for details.
+    StartCounters {
+        /// Counter index base.
+        counter_index_base: u64,
+        /// Counter index mask.
+        counter_index_mask: u64,
+        /// Counter start flags.
+        start_flags: PmuCounterStartFlags,
+        /// Counter initial value (used in conjunction with start_flags).
+        initial_value: u64,
+    },
+    /// Stops the counters selected by counter_index and counter_index_mask.
+    /// See the sbi_pmu_counter_stop documentation for details.
+    StopCounters {
+        /// Counter index base.
+        counter_index_base: u64,
+        /// Counter index mask.
+        counter_index_mask: u64,
+        /// Counter stop flags.
+        stop_flags: PmuCounterStopFlags,
+    },
+    /// Returns the current value firmware counter specified by the inner value.
+    ReadFirmwareCounter(u64),
+}
+
+/// This encapsulates the bit-fields for PMU config_flags parameter as described in the SBI documentation
+/// for sbi_pmu_counter_config_matching
+#[derive(Copy, Clone, Default)]
+pub struct PmuCounterConfigFlags(u64);
+
+impl PmuCounterConfigFlags {
+    /// Constructs a new PmuCounterConfigFlags from a valid passed-in value.
+    pub fn from_regs(value: u64) -> Result<Self> {
```

Comment:
nit: regs doesn't seem like it should be plural with one u64 arg.

Reply:
Seems like making it one-off will make it inconsistent with other functions -- no?

Reply:
Got...done.